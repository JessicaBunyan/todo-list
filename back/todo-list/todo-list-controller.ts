import { TodoList } from "./todo-list.model";
import { Router, Request, Response } from "express";

const startupState: { [key: string]: TodoList } = {
  a: {
    id: "a",
    name: "Shopping",
    items: [
      {
        id: "z",
        text: "Apples",
        done: false,
      },
      {
        id: "y",
        text: "Bananas",
        done: false,
      },
      {
        id: "x",
        text: "Carrots",
        done: true,
      },
    ],
  },
  b: {
    id: "b",
    name: "Chores",
    items: [
      {
        id: "p",
        text: "Hoovering",
        done: true,
      },
      {
        id: "q",
        text: "Dusting",
        done: false,
      },
      {
        id: "o",
        text: "Gardening",
        done: true,
      },
    ],
  },
  c: {
    id: "c",
    name: "Chores copy",
    items: [
      {
        id: "p",
        text: "Hoovering",
        done: true,
      },
      {
        id: "q",
        text: "Dusting",
        done: false,
      },
      {
        id: "o",
        text: "Gardening",
        done: true,
      },
    ],
  },
  d: {
    id: "d",
    name: "Chores copy 2",
    items: [
      {
        id: "p",
        text: "Hoovering",
        done: true,
      },
      {
        id: "q",
        text: "Dusting",
        done: false,
      },
      {
        id: "o",
        text: "Gardening",
        done: true,
      },
    ],
  },
};

const lists: { [key: string]: TodoList } = startupState;

export const todoListRoutes = function (router: Router) {
  router.post("/list", saveList);
  router.get("/list/all", getAllLists);
  router.get("/list/:id", getList);
  router.patch("/list/:listId/:itemId/:state", updateItem);
};

async function saveList(req: Request, res: Response) {
  const list: TodoList = req.body;

  list.id = list.id || getId();
  lists[list.id] = req.body;

  return res.json(list);
}

async function getAllLists(req: Request, res: Response) {
  console.log("in get all lists");
  res.json(
    Object.values(lists).map((l) => ({
      id: l.id,
      name: l.name,
      numItems: l.items.length,
    })),
  );
}

async function getList(req: Request, res: Response) {
  console.log("in get list");
  const id = req.params.id;

  const list = lists[id];
  if (!list) {
    return res.status(404).send("List not found");
  }

  res.json(list);
}

async function updateItem(req: Request, res: Response) {
  const { listId, itemId, state } = req.params;

  const list = lists[listId];
  if (!list) {
    return res.status(400).json("Invalid update - list does not exist");
  }
  const item = list.items.find((i) => i.id == itemId);

  if (!item) {
    return res.status(400).send("Invalid update - list item does not exist");
  }

  item.done = state == "true";

  res.send(list);
}

function getId() {
  return Math.random().toString(36).substring(2, 9);
}
