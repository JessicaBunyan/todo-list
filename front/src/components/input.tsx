import * as React from "react";

export interface ISubmittableInputProps {
  placeholder?: string;
  onSubmit: (s: string) => void;
}

export function SubmittableInput(props: ISubmittableInputProps) {
  const [val, setVal] = React.useState("");

  const onKeyPress = (ev: React.KeyboardEvent<HTMLInputElement>) => {
    console.log(ev);
    console.log(ev.currentTarget.value);
    if (ev.key == "Enter" && val) {
      props.onSubmit(val);
      setVal("");
      //   ev.currentTarget.value = "";
    }
  };

  const onClick = () => {
    if (val) {
      props.onSubmit(val);
      setVal("");
    }
  };

  return (
    <div className="input-group">
      <input
        type="text"
        className="form-control"
        placeholder={props.placeholder || ""}
        value={val}
        onKeyPress={onKeyPress}
        onChange={(e) => setVal(e.target.value)}
      />
      <div className="input-group-append">
        <button onClick={onClick} className="btn btn-outline-secondary">
          Add
        </button>
      </div>
    </div>
  );
}
