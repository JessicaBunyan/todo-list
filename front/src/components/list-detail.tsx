import * as React from "react";
import { ITodoList } from "../models/todo-list.model";
import { ITodoListDetail } from "./../models/todo-list.model";
import { Link } from "react-router-dom";

export interface IListCardProps {
  detail: ITodoListDetail;
}

export function TodoListDetail(props: IListCardProps) {
  return (
    <Link to={"/list/" + props.detail.id}>
      <div className="card">
        <h3 className="">
          <p>
            {props.detail.name}
            <span className="float-right">{props.detail.numItems}</span>
          </p>
        </h3>
      </div>
    </Link>
  );
}
