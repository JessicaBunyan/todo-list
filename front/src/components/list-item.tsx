import * as React from "react";
import { ITodoListItem, ITodoList } from "../models/todo-list.model";
import { patch } from "../utils/http";
import { useDispatch } from "react-redux";
import { setActiveList } from "../actions/todo-list-actions";

export interface ITodoListItemProps {
  index: number;
  listId: string;
  item: ITodoListItem;
}

export function TodoListItem(props: ITodoListItemProps) {
  const { item, index } = props;
  const dispatch = useDispatch();

  function onChange(itemId: string) {
    patch(
      `http://localhost:8080/api/list/${props.listId}/${itemId}/${!item.done}`,
    ).then((list: ITodoList) => {
      dispatch(setActiveList(list));
    });
  }

  return (
    <div className="item ">
      <span className={item.done ? "strikethrough" : ""}>
        {index + 1}. {item.text}
      </span>
      <input
        className="float-right"
        type="checkbox"
        checked={item.done}
        onChange={() => onChange(item.id)}
      ></input>
    </div>
  );
}
