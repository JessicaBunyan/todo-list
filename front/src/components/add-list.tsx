import * as React from "react";
import { SubmittableInput } from "./input";
import { post } from "../utils/http";
import { ITodoList } from "../models/todo-list.model";
import { useHistory } from "react-router";
import { ITodoListItem } from "./../models/todo-list.model";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faPlus } from "@fortawesome/free-solid-svg-icons";

export interface IAddListProps {}

export function AddList(props: IAddListProps) {
  const [active, setActive] = React.useState(false);
  const history = useHistory();

  const onSubmit = (val: string) => {
    const list: ITodoList = {
      name: val,
      items: [],
    };
    post("http://localhost:8080/api/list", list).then((l: ITodoList) => {
      history.push("/list/" + l.id);
    });
  };
  return (
    <div className="col-sm-12 col-md-6 col-lg-4 mb-4">
      <div
        className={active ? "card" : "card clickable"}
        onClick={() => setActive(true)}
      >
        {active && (
          <div className="mt-6 clickable">
            <SubmittableInput
              placeholder="New List Name..."
              onSubmit={onSubmit}
            ></SubmittableInput>
          </div>
        )}

        {!active && (
          <div className="text-center mt-4">
            <h3>New</h3>
            <FontAwesomeIcon className="mega" icon={faPlus}></FontAwesomeIcon>
          </div>
        )}
      </div>
    </div>
  );
}
