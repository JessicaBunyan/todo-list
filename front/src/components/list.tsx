import * as React from "react";
import { TodoListItem } from "./list-item";
import { get, post } from "./../utils/http";
import { ITodoList } from "./../models/todo-list.model";
import { useDispatch } from "react-redux";
import { setActiveList } from "../actions/todo-list-actions";
import { useTypedSelector } from "./../use-typed-selector";
import { useParams } from "react-router";
import { Link } from "react-router-dom";
import { SubmittableInput } from "./input";

export interface ITodoListProps {
  list: ITodoList;
}

export function TodoList(props: ITodoListProps) {
  console.log("in todo list");
  const dispatch = useDispatch();
  const { listId } = useParams();

  React.useEffect(() => {
    get("http://localhost:8080/api/list/" + listId).then((list: ITodoList) => {
      dispatch(setActiveList(list));
    });
  }, []);

  const list = useTypedSelector((s) => s.todoList && s.todoList.list);

  if (!list) {
    return null;
  }

  const addItem = (item: string) => {
    list.items.push({
      id: "" + Math.random() * 10000,
      text: item,
      done: false,
    });
    post("http://localhost:8080/api/list", list).then((list: ITodoList) => {
      dispatch(setActiveList(list));
    });
  };

  return (
    <div className="">
      <Link to="/">
        <button className=" text-center btn-lg btn-light">Home</button>
      </Link>
      <div className="list card">
        <h3>{list.name}</h3>
        {list.items.map((item, index) => {
          return (
            <TodoListItem
              key={item.id}
              item={item}
              listId={list.id}
              index={index}
            />
          );
        })}
        <SubmittableInput
          placeholder={"Add another item... "}
          onSubmit={addItem}
        ></SubmittableInput>
      </div>
    </div>
  );
}
