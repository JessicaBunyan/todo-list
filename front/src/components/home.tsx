import * as React from "react";
import { get } from "./../utils/http";
import { ITodoList } from "./../models/todo-list.model";
import { useTypedSelector } from "./../use-typed-selector";
import { setAllLists } from "../actions/todo-list-actions";
import { useDispatch } from "react-redux";
import { TodoList } from "./list";
import { TodoListDetail } from "./list-detail";

export interface IHomeProps {}

export function Home(props: IHomeProps) {
  const lists = useTypedSelector((s) => s.todoList && s.todoList.listDetails);
  const dispatch = useDispatch();
  console.log(lists);
  React.useEffect(() => {
    get("http://localhost:8080/api/list/all").then((lists: ITodoList[]) => {
      dispatch(setAllLists(lists));
      console.log(lists);
    });
  }, []);

  return (
    <div>
      <div className="row">
        {lists &&
          lists.map((l) => (
            <div key={l.id} className="col-sm-12 col-md-6 col-lg-4 mb-4">
              <TodoListDetail detail={l} />
            </div>
          ))}
      </div>
    </div>
  );
}
