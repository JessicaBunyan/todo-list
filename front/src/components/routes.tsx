import { BrowserRouter as Router, Route, Switch } from "react-router-dom";
import * as React from "react";
import App from "./../App";
import { TodoList } from "./list";
import { Home } from "./home";

export function Routes() {
  return (
    <Router>
      <Switch>
        <Route path="list/:listId" component={TodoList}></Route>
        <Route path="lists"></Route>
        <Route path="/" component={Home} />
      </Switch>
    </Router>
  );
}
