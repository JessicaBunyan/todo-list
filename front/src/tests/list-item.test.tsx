jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useDispatch: jest.fn(),
}));

// jest.mock("react-router", () => ({
//   useParams: jest.fn().mockReturnValue({ id: "123" }),
// }));
// jest.mock("react-router-dom", () => ({
//   useParams: jest.fn().mockReturnValue({ id: "123" }),
// }));

import React from "react";
import renderer from "react-test-renderer";
import { TodoListItem } from "../components/list-item";
import { ITodoList } from "../models/todo-list.model";
import { Provider } from "react-redux";
import { BrowserRouter as Router } from "react-router-dom";

const testList: ITodoList = {
  id: "foo",
  name: "Jessica test",
  items: [
    { id: "fsd", text: "this is an item", done: false },
    { id: "ffdsfds", text: "this is also an item", done: false },
  ],
};

it("renders correctly when checked", () => {
  const tree = renderer
    .create(
      <TodoListItem item={testList.items[0]} listId={testList.id} index={0} />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
