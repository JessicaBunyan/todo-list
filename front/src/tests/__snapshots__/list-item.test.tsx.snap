// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when checked 1`] = `
<li
  className="item pr-2"
>
  <span
    className=""
  >
    1
    . 
    this is an item
  </span>
  <input
    checked={false}
    className="float-right px-4 checkbox"
    onChange={[Function]}
    type="checkbox"
  />
</li>
`;
